### ISSUE TITLE
Notebook: edge case consistency

### ISSUE THREAD

**Issue by martroben** [Bertverbeek4PS/bc2adls#201]
I might have spotted a peculiarity in the [Fabric notebook code](https://github.com/Bertverbeek4PS/bc2adls/blob/main/fabric/CopyBusinessCentral.ipynb) which could lead to divergences in Business Central and OneLake data.

Suppose we have the following existing data in OneLake:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| existing value |    1     |   2020-01-01    |   2020-01-01     |
+----------------+----------+-----------------+------------------+
```

We're using the notebook to process two incoming delta csv-s:
```
9FBE22D6-... .csv
Date Modified:  2024-12-09
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
+----------------+----------+-----------------+------------------+

07E2D6DC-... .csv
Date Modified:  2024-12-10
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
|                |    1     |                 |                  |
+----------------+----------+-----------------+------------------+
```

Since `glob.glob()` returns results in [arbitrary order](https://docs.python.org/3/library/glob.html) (which looks alphabetical in case of OneLake), the notebook would process the deletion before the modification. That would result in OneLake having a record that the corresponding Business Central table does not have.

Is there something that I have overlooked or is it possible that something like this could happen in the current implementation? It is admittedly a low probability edge case, but an exceedingly difficult one to trace, should it come to pass.

**Comment by Bertverbeek4PS** [comment]
Thanks @martroben for the question. 
In my opinion it doesn't matter how the notebook is processing the files.
He will put all the new files in a dataframe together with the excisting data in OneLake.
Together then you got one dataframe with the old records and newly records (including the deleted records).

Then the script will remove all the deleted records first (together will all the records that are there).

After that he will remove the duplicates and keeping the newest once.

So he first will delete the deleted records based by systemID:
`
#delete all old records
            df_deletes = df_new.filter(df_new['SystemCreatedAt-2000000001'].isNull())
            if ContainsCompany:
                df_new = df_new.join(df_deletes, ['$Company','systemId-2000000000'], 'leftanti')
            else:
                df_new = df_new.join(df_deletes, ['systemId-2000000000'], 'leftanti')
`

So in my opinion it will process everything in a right way.

**Comment by martroben** [comment]
@Bertverbeek4PS, let me try to explain my logic using the above example to see where I'm going wrong.
### Delta files to be processed
`glob.glob(folder_path + entry.name + '/*')` returns the following paths. We're going to have two iterations in the for loop: 
```
[
	'/lakehouse/default/Files/deltas/object/07E2D6DC-... .csv',
	'/lakehouse/default/Files/deltas/object/9FBE22D6-... .csv',
]
```
### TL;DR
First iteration saves an empty data frame, second iteration appends a deleted row.

---
### 1st iteration: `07E2D6DC-... .csv`
#### Filename
`filename = '/lakehouse/default/Files/deltas/object/07E2D6DC-... .csv'`
#### Incoming delta data
`df_new = spark.read.option("minPartitions", no_Partition).format("csv").option("header","true").load(folder_path_spark + entry.name +"/*")` loads both csv-s:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
|                |    1     |                 |                  |
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
+----------------+----------+-----------------+------------------+
```
#### File list
`file_list.append(filename)` assigns following value to the `file_list` variable:
`['/lakehouse/default/Files/deltas/object/07E2D6DC-... .csv']` 
#### Existing data
`df_old = spark.sql(SQL_Query)` gives:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| existing value |    1     |   2020-01-01    |   2020-01-01     |
+----------------+----------+-----------------+------------------+
```
#### Combined data
`df_new = df_new.union(df_old).repartition(no_Partition)` gives:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
|                |    1     |                 |                  |
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
| existing value |    1     |   2020-01-01    |   2020-01-01     |
+----------------+----------+-----------------+------------------+
```
#### Deletes
`df_deletes = df_new.filter(df_new['SystemCreatedAt-2000000001'].isNull())` gives:

```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
|                |    1     |                 |                  |
+----------------+----------+-----------------+------------------+
```

Then `df_new = df_new.join(df_deletes, ['systemId-2000000000'], 'leftanti')` gives an empty data frame.
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
+----------------+----------+-----------------+------------------+
```
#### Data saved to lakehouse
`df_new.write.mode("overwrite").format("delta").save("Tables/" + table_name)` saves the empty data frame to the lakehouse.
#### Remaining delta files
The last bit of code deletes `07E2D6DC-... .csv` from the deltas folder.
```python
for filename in file_list:  
	try:  
		os.remove(filename)  
	except OSError as e:  # this would catch any error when trying to delete the file  
		print(f"Error: {filename} : {e.strerror}")
```

The `file_list` variable is reset in the end.
### 2nd iteration: `9FBE22D6-... .csv`
#### Filename
`filename = '/lakehouse/default/Files/deltas/object/9FBE22D6-... .csv'`
#### Incoming delta data
`df_new = spark.read.option("minPartitions", no_Partition).format("csv").option("header","true").load(folder_path_spark + entry.name +"/*")` loads the remaining csv:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
+----------------+----------+-----------------+------------------+
```
#### File list
`file_list.append(filename)` assigns following value to the `file_list` variable:
`['/lakehouse/default/Files/deltas/object/9FBE22D6-... .csv']` 
#### Existing data
`df_old = spark.sql(SQL_Query)` loads the empty data frame saved in the last iteration:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
+----------------+----------+-----------------+------------------+
```
#### Combined data
`df_new = df_new.union(df_old).repartition(no_Partition)` gives:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
+----------------+----------+-----------------+------------------+
```
#### Deletes
`df_deletes = df_new.filter(df_new['SystemCreatedAt-2000000001'].isNull())` gives and empty data frame, therefore `df_new = df_new.join(df_deletes, ['systemId-2000000000'], 'leftanti')` has no effect.

`df_new` still has the single row from the remaining csv:
```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
+----------------+----------+-----------------+------------------+
```
#### Data saved to lakehouse
`df_new.write.mode("overwrite").format("delta").save("Tables/" + table_name)` saves `df_new` to the lakehouse.

```
+----------------+----------+-----------------+------------------+
|      Text      | systemId | SystemCreatedAt | SystemModifiedAt |
+----------------+----------+-----------------+------------------+
| updated value  |    1     |   2020-01-01    |   2024-12-09     |
+----------------+----------+-----------------+------------------+
```
#### Remaining delta files
The last bit of code deletes `9FBE22D6-... .csv` from the deltas folder, completing the loop.

---
The end result is that the lakehouse table now contains a row that is actually deleted from Business Central.

**Comment by Bertverbeek4PS** [comment]
@martroben I have tested it. 
It was not an issue because it was taken all the three files in one dataframe.
But with remove_delta set to false it was doing this three times. 

So fixed that with the following pull request:
https://github.com/Bertverbeek4PS/bc2adls/pull/202
It iterates only on the directories and if the directory is empty then it will skip.

Thanks for your explaining it and your help!!

**Comment by martroben** [comment]
Yes, I think this will help to avoid the unintended effects that I was worried about.
Thank you for the quick reply and code fix.
